### PR TITLE
Don't serialize I/O aliases for FusionCache. 

### PR DESCRIPTION
**Pull request by wujingyue**
Instead, use `kernel->ioAlias()`. I believe this doesn't break anything: 
1. Aliases added when building `FusionDefinition`s (e.g. by calling `fd.ops.batch_norm` or `fd.add_output(..., alias_input=...)`) are replayed by `RecordFunctors`. 
2. Aliases added during scheduling or later (e.g. by `MarkAliasPass`) are re-added by optimization passes or schedulers.

We already have `test_alias_output_to_input` for case 1, and this PR added `test_mark_alias_pass` for case 2.

### CODE REVIEW

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
!build

**Review by jjsjann123 — COMMENTED** [review]
code looks fine.

But this is surprising to me. How is the alias restored after deserialization if we are not serializing it in the first place?!
I'm wondering if we have any test in place checking that we are indeed restoring the alias information. I think trivial forwarding *might* be ok?! i.e. we are outputting the same tensor anyway. But for the running stats in normalization, I don't remember any other place in fusion that marks it as in-place update...

**Comment by jjsjann123** [comment]
We should probably add a test for serialization on aliased inplace update.

The only inplace update I'm seeing is with `FusionSegmentIoAlias_CUDA` in test/test_gpu2.cpp.
We should be able to mimic that on python API as well. there's `add_output` that accepts an aliased input.

Though I don't know how to hook up serde in the test. cc'ing @rdspring1

**Comment by wujingyue** [comment]
It surprises me too. My current hypothesis (which I'll verify) is that, **currently**, aliases are added at the time of building the fusion definition (e.g. csrc/ops/normalization.cpp). Therefore, it's OK to not serialize/deserialize aliases because new fusion definitions will have them before the cache lookup. However, soon, I'll add an optimization pass that marks aliases, and this PR will break that (which means I should abandon this PR).

**Comment by rdspring1** [comment]
Serialization is only tested in `python_test/test_python_frontend.py`, so alias testing is covered by [test_alias_output_to_input](https://github.com/NVIDIA/Fuser/blob/main/python_tests/test_python_frontend.py#L710-L725).

Alias information is stored by the `OutputRecord` in the fusion definition. Serialization saves the `OutputRecord` for each fusion and rebuilds [the alias relationships](https://github.com/NVIDIA/Fuser/blob/main/csrc/python_frontend/fusion_record.h#L1417-L1425) during deserialization.

**Comment by rdspring1** [comment]
@jjsjann123 Maybe we should have a `batch_norm` test in our python frontend because a reference for it does not exist yet.

**Comment by wujingyue** [comment]
> Serialization is only tested in `python_test/test_python_frontend.py`, so alias testing is covered by [test_alias_output_to_input](https://github.com/NVIDIA/Fuser/blob/main/python_tests/test_python_frontend.py#L710-L725).
> 
> > How is the alias restored after deserialization if we are not serializing it in the first place?!
> 
> Alias information is stored by the `OutputRecord` in the fusion definition. Serialization saves the `OutputRecord` for each fusion and rebuilds [the alias relationships](https://github.com/NVIDIA/Fuser/blob/main/csrc/python_frontend/fusion_record.h#L1417-L1425) during deserialization.

The title of my PR is misleading. I don't think it changes the logic around `OutputRecord` or the Python-to-C++ bridge. Instead, it removes `input_output_aliases` from https://github.com/NVIDIA/Fuser/blob/73c4e31ae9ed826e914edba7062dd7e8cdc74a25/csrc/executor.h#L145. This is related to the kernel cache etc.

**Comment by wujingyue** [comment]
I'm closing this PR. I managed to find a way to break it with my pending #1203. Will add a regression test over there.

**Comment by wujingyue** [comment]
Actually, the test isn't fair because the first thing `self.exec_nvfuser` does is to **deepcopy** the inputs... cc @rdspring1

**Comment by wujingyue** [comment]
> Actually, the test isn't fair because the first thing `self.exec_nvfuser` does is to **deepcopy** the inputs... In fact, the PR itself looks good -- it produces an output of the same `data_ptr` as the **deepcopied** input. cc @rdspring1

OK -- I think I fixed the code around the deepcopy.

**Review by jjsjann123 — APPROVED** [review]
minor comment on tests

**Comment by jjsjann123** [comment]
!build

File: python_tests/test_python_frontend.py [hunk 1]
```diff
@@ -713,22 +714,22 @@ def test_alias_output_to_input(self):
...
-            t0 = fd.from_pytorch(inputs[0])
+            t0 = fd.from_pytorch(inputs[0])  # 1.0
             s0 = fd.define_scalar(1.0)
             s1 = fd.define_scalar(2.0)
             s2 = fd.define_scalar(3.0)
-            t1 = fd.ops.add(t0, s0)
-            t2 = fd.ops.add(t0, s1)
-            t3 = fd.ops.add(t2, s2)
+            t1 = fd.ops.add(t0, s0)  # 2.0
+            t2 = fd.ops.add(t0, s1)  # 3.0
```
[the remark this answers is not in the document]
Comment (wujingyue) [reply python_tests/test_python_frontend.py]: Done. I fixed the test to avoid this undefined behavior.

**Comment by jjsjann123** [comment]
:shipit:

**Pull request merged**